### ISSUE TITLE
Provide an option to configure the url pattern for RUM data

### ISSUE THREAD

**Issue by hmdhk** [elastic/apm-server#3868]
To address the [high url cardinality of RUM data](https://github.com/elastic/apm-agent-rum-js/issues/56) we decided as a first step to provide a configuration for the users to provide a url pattern that will help us group transactions based on the page url. 

This config could accept a regex or another syntax for defining the url pattern.

Furthermore, this option could be fetched from central config. This has a few advantages, for example it lets both Kibana and potentially the RUM agent to consider this pattern on grouping and sampling of transactions. As well as providing easy and scalable configuration for the users

For more context [see this comment](https://github.com/elastic/apm-agent-rum-js/issues/56#issuecomment-641913346)

cc @axw

**Comment by axw** [comment]
Thanks for opening this @jahtalab

We also discussed the possibility of using [ML categorisation](https://www.elastic.co/guide/en/machine-learning/7.7/ml-configuring-categories.html). I think these would work well together: automatic categorisation would [produce regular expressions](https://www.elastic.co/guide/en/elasticsearch/reference/7.7/ml-get-category.html) that the APM Server could periodically fetch, to apply to future transactions. Users could manually override these with their own rules, defined in the UI, for more control.

**Comment by axw** [comment]
@jahtalab @vigneshshanmugam given that we have https://github.com/elastic/apm-agent-rum-js/issues/56, do you still think this enhancement is worthwhile? Do we have any data on how well the heuristic-based approach is working?

**Comment by vigneshshanmugam** [comment]
@axw Cant say for now if our current approach is performing well enough, we have deployed the new version of the agent on our website. Would like to leave it for a couple of days to analyse the data points and can confirm the same. WDYT?

**Comment by axw** [comment]
Sure, there's no rush I'm just doing some spring cleaning :)

**Comment by sqren** [comment]
We are currently running into this issue with data from elastic.co:

![image](https://user-images.githubusercontent.com/209966/97422987-e197f180-190e-11eb-9f7d-05c2f32c19d7.png)

A quick cardinality count shows that there are 1905 unique transaction groups within the past week. That seems quite high. A look into the groups show they are duplicated for each language. Take for instance the subscriptions page:

![image](https://user-images.githubusercontent.com/209966/97423080-01c7b080-190f-11eb-8ee9-ddc6dd8a1537.png)

There's already some custom grouping happening in the RUM agent but it's clearly not a simple task. Would APM Server be in a better position to group related transactions?

**Comment by jalvz** [comment]
I'm inclined to say that we could do something about it. 
We have the infrastructure already to aggregate transactions for histogram metrics, so we would "just" have to plug in the extra bits to aggregate by a new criteria; and we probably can afford the extra computational cycles that RUM can't.

Not to say that it would be easy or we should commit to it, but I certainly would be happy to have a look.

**Comment by axw** [comment]
Agreed, we certainly could do something about it in the server – but it could be a bit dangerous.

Breakdown metrics, which are calculated by agents, contain the transaction name observed by the agent. Similarly, if we ever want to push the transaction duration histogram aggregation down to the agents (for higher accuracy), then we would need to aggregate on the transaction name.

That's not to say it's impossible, it's just something we'll need to be cautious about to ensure we don't end up with inconsistencies. The server would need to apply the renaming to both trace events and metrics, and potentially combine the aggregated metrics.

**Comment by simitt** [comment]
Could we introduce something like a `transaction.group_name` that by default is set to the `transaction.name` but can be used to group multiple similar `transaction.names` together, instead of changing the `transaction.name` directly to not lose information? The UI would then need to use the new field and potentially fall back to transaction name if not available.
